In [1]:
import torch
import numpy as np
import pandas as pd
import os

In [2]:
os.listdir()

['model_FCNet3.pt',
 'model_FCNet1.pt',
 'model_FCNet2.pt',
 'Untitled.ipynb',
 'networks_test_on_scVI_dataset_bk3.ipynb',
 '.ipynb_checkpoints',
 'cortex_raw_latent.csv',
 'networks_test_on_scVI_dataset.ipynb',
 'networks_test_on_scVI_dataset_test1.ipynb',
 'networks_test_on_scVI_dataset_bk2.ipynb',
 'metadata.xlsx']

In [3]:
not_readin_list = ['Zn1','Zn2','Zn3','Zn4','Zn5','Zn6','Zn7','Zn8','Zn9','Zn10']

scVI_data_raw = pd.read_csv('cortex_raw_latent.csv', sep=',',index_col=0, usecols=lambda x:x not in not_readin_list)

In [4]:
#change to dtype='float32' 
#  because the default dtype in torch.nn.Module uses 'float32'
scVI_data_raw=scVI_data_raw.astype('float32')

In [5]:
scVI_data_raw

SUB1  CXCL14   SCG2  GM9846  HTR3A   CNR1  GNG2  NRXN3  RPH3A  TAC2  \
0     28.0    22.0  116.0     7.0  131.0   20.0   5.0   22.0   21.0  82.0   
1     41.0    18.0   12.0    17.0  371.0   12.0   6.0   11.0   39.0  21.0   
2     57.0    26.0   46.0     9.0  300.0  122.0  26.0   87.0   10.0  28.0   
3     33.0    93.0   48.0     9.0  415.0   82.0  12.0   27.0   38.0  47.0   
4     30.0     0.0   67.0     3.0  131.0    7.0  13.0   26.0   13.0   0.0   
5     13.0    22.0   23.0     2.0  247.0   40.0  10.0   30.0   27.0  49.0   
6     27.0    29.0   54.0    13.0   46.0   38.0   9.0   57.0   50.0   0.0   
7     17.0     9.0   15.0     2.0  100.0   80.0   5.0   41.0   15.0   0.0   
8     23.0     2.0   77.0     8.0   35.0   10.0   3.0   34.0   24.0   0.0   
9     31.0     9.0   34.0    11.0   75.0   62.0  10.0   38.0   14.0   0.0   
10    34.0     1.0   52.0     4.0   56.0   91.0   8.0   25.0   18.0   2.0   
11    36.0     0.0   25.0    10.0    0.0   21.0  14.0   22.0   34.0   0.0   
12    21.0     0.0   23.0     9.0    0.0    7.0   7.0   39.0   17.0   0.0   
13    25.0     0.0   13.0     1.0    2.0   18.0  15.0   38.0    3.0   0.0   
14    25.0     0.0   77.0     3.0    1.0    3.0   8.0   83.0   26.0   0.0   
15    29.0     0.0   63.0     7.0    0.0    3.0  22.0   74.0   37.0   0.0   
16    44.0     0.0   35.0     2.0   27.0    2.0  18.0   32.0    0.0   0.0   
17    28.0     0.0   30.0     1.0   15.0    1.0  12.0   85.0    2.0   0.0   
18    70.0     0.0    7.0     1.0    0.0    2.0   4.0   22.0    5.0   0.0   
19    20.0     0.0    7.0     1.0    0.0    1.0   8.0    4.0    9.0   0.0   
20    51.0     0.0   39.0     5.0   13.0    2.0  35.0    7.0   24.0   0.0   
21    23.0     0.0   27.0     4.0    0.0    0.0   9.0    4.0   11.0   0.0   
22    36.0     0.0   33.0    17.0   84.0   50.0   0.0   18.0   18.0   0.0   
23    40.0     1.0    8.0     8.0    0.0   16.0  10.0    6.0   11.0   0.0   
24    52.0     0.0   22.0    11.0    5.0    8.0  12.0   20.0   12.0   0.0   
25    50.0     8.0   77.0    13.0   55.0   42.0   8.0   38.0   15.0   0.0   
26    51.0    10.0   40.0     8.0   28.0   11.0   9.0   22.0   49.0   0.0   
27    39.0     0.0   23.0     8.0    7.0    0.0   0.0    2.0    3.0   0.0   
28    30.0     0.0   51.0     4.0   12.0    4.0   9.0    6.0    8.0   0.0   
29    52.0     0.0   56.0     8.0    6.0    1.0   8.0    7.0   11.0   0.0   
...    ...     ...    ...     ...    ...    ...   ...    ...    ...   ...   
2975   0.0     0.0    1.0     0.0    0.0    0.0   0.0    1.0    0.0   0.0   
2976   3.0     0.0   12.0     0.0    0.0    0.0   1.0    2.0    0.0   0.0   
2977   2.0     0.0    0.0     8.0    0.0    0.0   0.0    0.0    0.0   0.0   
2978   0.0     0.0    0.0     3.0    0.0    0.0   0.0    0.0    0.0   0.0   
2979   0.0     0.0    0.0     3.0    0.0    0.0   0.0    2.0    3.0   0.0   
2980   4.0     0.0    0.0     3.0    0.0    0.0   0.0    0.0    0.0   0.0   
2981   1.0     0.0    7.0     5.0    0.0    0.0   0.0    0.0    2.0   0.0   
2982   1.0     0.0    5.0     4.0    2.0    0.0   0.0    0.0    6.0   0.0   
2983   3.0     0.0    0.0     1.0    0.0    0.0   0.0    3.0    0.0   0.0   
2984   1.0     0.0    6.0     0.0    0.0    0.0   0.0    0.0    0.0   0.0   
2985   0.0     0.0    0.0     5.0    0.0    0.0   0.0    0.0    0.0   0.0   
2986   7.0     0.0    0.0     4.0    0.0    0.0   2.0    1.0    0.0   0.0   
2987   1.0     1.0    0.0     5.0    0.0    0.0   0.0    1.0    0.0   0.0   
2988   3.0     0.0    0.0     1.0    0.0    0.0   0.0    0.0    0.0   0.0   
2989   2.0     0.0    1.0     2.0    0.0    0.0   0.0    5.0    1.0   0.0   
2990   3.0     0.0    0.0    10.0    0.0    0.0   2.0    0.0    0.0   0.0   
2991   1.0     0.0    0.0     1.0    0.0    0.0   0.0    0.0    0.0   0.0   
2992  16.0     6.0   36.0     1.0    0.0   11.0   4.0    2.0    2.0  19.0   
2993   0.0     0.0    0.0     8.0    0.0    0.0   0.0    0.0    0.0   0.0   
2994   3.0     0.0    0.0     2.0    0.0    0.0   0.0    2.0    0.0   0.0   
29

In [6]:
columns_names = scVI_data_raw.columns.values
features = columns_names[:-1]
print(features, f'length of features is {len(features)}')

['SUB1' 'CXCL14' 'SCG2' 'GM9846' 'HTR3A' 'CNR1' 'GNG2' 'NRXN3' 'RPH3A'
 'TAC2' 'XIST' 'NAP1L5' 'MARCKS' 'PEBP1' 'GNAS' 'KIF1A' 'PGAM1' 'TCF25'
 'STXBP1' 'TSPYL4' 'PRKAR1B' 'ALDOA' 'PKM' 'GAPDH' 'ATP6V1B2' 'MLLT11'
 'CPLX1' 'SLC25A3' 'PCLO' 'HSPA8' 'NDN' 'RPL41' 'NTRK2' 'SYT1' 'GABARAPL1'
 'NDRG4' 'SV2A' 'RAB3A' 'IMPACT' 'CELF4' 'GM6682' 'SYP' 'OXCT1'
 'ATP6V0C-PS2' 'VAMP2' 'RAB6B' 'PURB' 'ATP6V1A' 'YWHAG' 'EIF4G2' 'UBC'
 'SNURF' 'CLTC' 'CACNB4' 'TSPAN3' 'APLP2' 'LARS2' 'NRIP3' 'SLC6A1'
 'OGFRL1' 'GAD1' 'GAD2' 'NRSN1' 'DNER' 'GABRB2' 'TSPAN7' 'GABRA1' 'RELN'
 'NPY' 'RESP18' 'CRHBP' 'SST' 'ZWINT' 'STMN3' 'NDUFA4' 'SERPINI1' 'SUMO1'
 'RPL7' 'SYNPR' 'NDUFV3' 'FXYD6' 'GM15421' 'NDUFA5' 'COX7B' 'RNASEK'
 'UQCRH' 'NDUFB10' 'EIF4A2' 'ATP5J2' 'LDHB' 'ATP5F1' 'NDUFA13' 'NDUFV2'
 'ATP5J' 'FKBP3' 'NDUFA11' 'NDUFA6' 'NDUFB9' 'NGFRAP1' 'AKR1A1' 'MDH1'
 'GPX4' 'COX4I1' 'NDUFB6' 'MYL12B' 'IDH3B' 'HINT1' 'GPI1' 'ATP6V1E1'
 'UCHL1' 'ATP5H' 'GM5506_LOC1' 'GABARAPL2' 'ATP5G3' 'ACOT7' 'GOT1'
 'TUBA1B' 'SLC

In [7]:
labels = scVI_data_raw['cell_labels'].unique()
print(labels)

[2. 6. 5. 4. 3. 1. 0.]


In [8]:
labels.sort()
labels

array([0., 1., 2., 3., 4., 5., 6.])

In [9]:
frequency = scVI_data_raw['cell_labels'].value_counts()

In [10]:
frequency.sort_index(inplace=True)
print(frequency)

0.0    224
1.0    235
2.0    290
3.0     98
4.0    820
5.0    939
6.0    399
Name: cell_labels, dtype: int64


In [11]:
target_weights = frequency.max()/frequency.values
print(target_weights)

[4.19196429 3.99574468 3.23793103 9.58163265 1.14512195 1.
 2.35338346]


In [12]:
#calculate the mean and std for each feature
norm_mean = scVI_data_raw.mean(axis=0)
norm_mean = norm_mean.values[:-1]
norm_std = scVI_data_raw.std(axis=0)
norm_std = norm_std.values[:-1]

In [13]:
num_features = norm_mean.size

In [14]:
norm_mean

array([1.94748745e+01, 1.54542434e+00, 6.39600658e+00, 4.81564045e+00,
       5.17637253e+00, 7.08419323e+00, 3.13976717e+00, 6.08918476e+00,
       2.17637277e+00, 7.63394356e-01, 7.26722145e+00, 5.26855230e+00,
       4.06955051e+00, 5.20266247e+00, 1.39480867e+01, 9.16472530e+00,
       6.27787018e+00, 6.37637281e+00, 4.66622305e+00, 3.78602338e+00,
       5.54908466e+00, 2.33514137e+01, 6.53643942e+00, 3.42163062e+00,
       5.87720490e+00, 6.31347752e+00, 3.44692183e+00, 7.32778692e+00,
       3.94542432e+00, 1.79717140e+01, 3.73943424e+00, 5.67154741e+00,
       4.91946745e+00, 2.45331116e+01, 5.48019981e+00, 1.09357738e+01,
       4.77204657e+00, 6.15640593e+00, 4.14742088e+00, 8.67620659e+00,
       5.05757093e+00, 4.95773697e+00, 5.24625635e+00, 1.67613983e+01,
       4.06455898e+00, 5.11314487e+00, 5.65424299e+00, 1.67524128e+01,
       1.04069881e+01, 5.17537451e+00, 7.98768711e+00, 1.13031616e+01,
       5.81963396e+00, 3.94908476e+00, 7.25590706e+00, 5.80765390e+00,
      

In [15]:
norm_std

array([ 19.996332 ,   8.0768   ,  13.455291 ,   9.971496 ,  30.343914 ,
        35.027294 ,   6.4459057,  12.570627 ,   5.5764427,   6.887598 ,
        11.024884 ,  12.002999 ,  11.5527   ,   6.647176 ,  18.92495  ,
         9.160055 ,   7.951228 ,   6.900118 ,   7.2170143,   5.740604 ,
         7.7904906,  26.477642 ,   8.191883 ,   7.3156614,   7.44464  ,
         8.135968 ,   7.987523 ,   8.456635 ,   6.7879977,  20.669798 ,
         6.479687 ,   6.325261 ,   5.9884076,  31.099289 ,   6.149114 ,
        15.274307 ,   6.502198 ,   8.073367 ,   5.717486 ,  11.53666  ,
         6.3752766,   6.6774564,   5.9962196,  19.795004 ,   5.5387354,
         6.9360194,   6.2472835,  18.475086 ,  12.779827 ,   5.4398594,
         8.528377 ,  14.251297 ,   6.696886 ,   5.512883 ,   6.963243 ,
         6.622773 ,  15.5763035,   7.5277634,  10.012262 ,   5.5365553,
        16.784956 ,  14.816611 ,   5.448512 ,   8.467054 ,   8.232136 ,
         9.318952 ,  11.904029 ,  10.020622 , 107.802376 ,   6.0

In [16]:
print("norm_mean's max:",norm_mean.max(), ", min:", norm_mean.min())
print("norm_std's max:",norm_std.max(), ", min:", norm_std.min())

norm_mean's max: 732.1524 , min: 0.5377704
norm_std's max: 648.1422 , min: 5.4398594


In [17]:
import sklearn
from sklearn.model_selection import train_test_split

In [18]:
data_train, data_valid= train_test_split(scVI_data_raw, test_size=0.2, random_state = 0)

In [19]:
data_train

SUB1  CXCL14  SCG2  GM9846  HTR3A   CNR1  GNG2  NRXN3  RPH3A  TAC2  ...  \
899   84.0     0.0  29.0     0.0    0.0    1.0  25.0    5.0    0.0   0.0  ...   
852   10.0     0.0   7.0     0.0    0.0    3.0   0.0   13.0    0.0   0.0  ...   
135   35.0     0.0   5.0     0.0    0.0    0.0   1.0    3.0    6.0   0.0  ...   
64    36.0    65.0  89.0     3.0  335.0  108.0  26.0   40.0   11.0  58.0  ...   
1691   1.0     0.0   0.0     3.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
2365   0.0     0.0   0.0     0.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
890   56.0     4.0   1.0     2.0    0.0    1.0   3.0    0.0    0.0   0.0  ...   
1380  40.0     0.0   1.0     0.0    0.0    2.0   3.0    2.0    0.0   0.0  ...   
675   36.0     0.0  12.0     1.0    0.0    0.0   6.0   12.0    4.0   0.0  ...   
1042  27.0     0.0  17.0     1.0    0.0    1.0   6.0    9.0    0.0   0.0  ...   
476   11.0     0.0   1.0     1.0    0.0    0.0   4.0    4.0   10.0   0.0  ...   
1379  34.0     0.0   5.0     0.0    0.0    1.0   2.0    4.0    0.0   0.0  ...   
2367  13.0     0.0   1.0    10.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
1324  64.0     0.0   3.0     1.0    0.0    1.0  12.0    1.0    0.0   0.0  ...   
745   19.0     0.0   2.0     6.0    0.0    1.0   8.0   10.0   16.0   0.0  ...   
271   20.0     0.0   6.0     2.0    0.0    2.0   1.0    3.0    2.0   0.0  ...   
1084  32.0     0.0  12.0     0.0    0.0    3.0   1.0   14.0    0.0   0.0  ...   
369   17.0     0.0   0.0     0.0    0.0    0.0   8.0    8.0    8.0   0.0  ...   
923   41.0     0.0   1.0     2.0    0.0    1.0   1.0    0.0    0.0   0.0  ...   
1231  46.0     0.0   1.0     1.0    0.0    8.0   9.0    7.0    0.0   0.0  ...   
1519  16.0     0.0   5.0     5.0    0.0    0.0   1.0    1.0    0.0   0.0  ...   
2885   1.0     0.0   0.0     0.0    0.0    1.0   0.0    0.0    0.0   0.0  ...   
1147  25.0     0.0  14.0     2.0    0.0    0.0   4.0    7.0    0.0   0.0  ...   
1095  52.0     0.0   4.0     6.0    0.0    3.0   7.0    6.0    0.0   0.0  ...   
687   10.0     0.0   2.0     3.0    0.0    0.0   0.0    5.0    0.0   0.0  ...   
109   62.0     0.0  49.0     8.0    0.0    0.0  13.0   33.0    7.0   0.0  ...   
728   26.0     0.0   0.0    12.0    0.0    1.0   1.0   13.0   11.0   0.0  ...   
2835   1.0     1.0   0.0     1.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
1269  57.0     0.0  10.0     1.0    0.0    4.0   9.0    9.0    0.0   0.0  ...   
793   19.0     0.0  46.0     0.0    0.0    5.0   0.0    3.0    0.0   0.0  ...   
...    ...     ...   ...     ...    ...    ...   ...    ...    ...   ...  ...   
2496   1.0     0.0   0.0     5.0    0.0    0.0   1.0    0.0    0.0   0.0  ...   
1871  11.0     0.0   0.0    16.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
2046   3.0     0.0   0.0     3.0    0.0    0.0   0.0    1.0    0.0   0.0  ...   
755    2.0     0.0  12.0     0.0    0.0    0.0   0.0   17.0    0.0   0.0  ...   
976   60.0     0.0   5.0     3.0    0.0    8.0   8.0   17.0    0.0   0.0  ...   
2163   1.0     0.0   0.0     1.0    0.0    0.0   0.0    1.0    2.0   0.0  ...   
1940   8.0     0.0   0.0     6.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
2825   1.0     1.0   1.0     1.0    0.0    1.0   0.0    0.0    1.0   0.0  ...   
2120  18.0     0.0   0.0     2.0    0.0    0.0   1.0    1.0    0.0   0.0  ...   
2893   4.0     0.0   1.0     1.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
537   23.0     1.0   1.0     2.0    0.0    0.0   3.0    3.0    5.0   0.0  ...   
1701   5.0     1.0   0.0     2.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
2897   5.0     1.0   6.0     1.0    0.0    0.0   0.0    3.0    0.0   0.0  ...   
2222   4.0     0.0   0.0     2.0    0.0    0.0   0.0    1.0    0.0   0.0  ...   
2135   3.0     0.0   0.0     0.0    0.0    0.0   0.0    1.0    0.0   0.0  ...   
2599   1.0     0.0   0.0    18.0    0.0    0.0   0.0    0.0    0.0   0.0  ...   
705   35.0     0.0  16.0     6.0    0.0    0.0   1.0    3.0    0.0   0.0  ...   
2362   3.0     0.0   0.0     6.0    0

In [20]:
type(data_train)

pandas.core.frame.DataFrame

In [21]:
#This is the Dataset class. It combined with Dataloader can be helpful in training
from torchvision import transforms#, utils
from torch.utils.data import Dataset, DataLoader
#class ScVIDataset can accept transforms that can transform the data
class ScVIDataset(Dataset):
    '''scVI dataset, 
        it will output {'features':features, 'target':target}
        it will take an optional argument 'transform' so that any required processing can be applied on the sample.
    '''
    def __init__(self, data_frame, target_col='cell_labels', transform=None):
        '''
        Args:
            data_frame(pandas.core.frame.DataFrame): It contains the dataframe
            target_col(string, default is 'cell_labels'): the target label column's name. default is 'cell_labels'
            transform(callable, optional): The transform that can be applied on a sample
        '''
        self.data_frame = data_frame.drop([target_col],axis=1)
        self.target = data_frame[target_col]
        self.transform = transform
                
    def __len__(self):
        return len(self.data_frame)
    
    def __getitem__(self, idx):
        #target = np.zeros(len(self.target_map))##DEL
        #target[self.target_map[self.target.iloc[idx]]] = 1##DEL
        sample = {'features':  self.data_frame.iloc[idx], 'target':self.target.iloc[idx].astype('int64')}
        if self.transform:
            sample = self.transform(sample)
        
        return sample


In [22]:
scVI_dataset = ScVIDataset(data_train, target_col='cell_labels')
sample = scVI_dataset[0]

In [23]:
sample

{'features': SUB1         84.0
 CXCL14        0.0
 SCG2         29.0
 GM9846        0.0
 HTR3A         0.0
 CNR1          1.0
 GNG2         25.0
 NRXN3         5.0
 RPH3A         0.0
 TAC2          0.0
 XIST          7.0
 NAP1L5        6.0
 MARCKS        1.0
 PEBP1        33.0
 GNAS         22.0
 KIF1A        42.0
 PGAM1         8.0
 TCF25         4.0
 STXBP1       20.0
 TSPYL4        3.0
 PRKAR1B      11.0
 ALDOA        83.0
 PKM           6.0
 GAPDH         3.0
 ATP6V1B2      6.0
 MLLT11       19.0
 CPLX1        12.0
 SLC25A3      34.0
 PCLO          3.0
 HSPA8        19.0
              ... 
 SRP14        13.0
 EIF2S2        5.0
 SEC62         6.0
 RPS3A1        6.0
 DNAJA1        8.0
 VCP           9.0
 RPS6          2.0
 ATP6V0B      13.0
 CDC42        12.0
 GM13826       4.0
 ARPC1A        8.0
 HMGB1        16.0
 SHFM1         3.0
 HNRNPA2B1    18.0
 CNBP         23.0
 GM6654        4.0
 NDUFA3        1.0
 RPS16         3.0
 RPL13A       10.0
 RPS3          0.0
 RPS13         4.0


In [24]:
type(sample['features'])

pandas.core.series.Series

In [25]:
sample['target']

5

In [26]:
type(sample['target'])

numpy.int64

In [27]:
#define the transforms we should use here:
#Normalize: to normalize to given mean and std
#ToTensor: to convert the pandas data frame to torch tensors
class Normalize(object):
    '''
    This is (feature - mean)/std for each feature in features
    Args:
        means(numpy.ndarray or python list): a list of means
        stds(numpy.ndarray or python list): a list of stds
    '''
    
    def __init__(self, means, stds):
        self.means = means
        self.stds = stds
    
    def __call__(self, sample):
        return {'features': (sample['features']-self.means)/self.stds, 'target': sample['target']}
    
class ToTensor(object):
    '''Convert the data frames in sample to Tensors'''
    def __call__(self, sample):
        #torch image: C X H X W
        feature = sample['features'].values # the shape is (#,), So, I need to expand its dimension
        return {'features': torch.from_numpy(feature[None,None,:]), 
            'target': sample['target']}#torch.from_numpy(sample['target'])}



In [28]:
composed_transforms = transforms.Compose([Normalize(norm_mean, norm_mean),
                                         ToTensor()])
scVI_dataset = ScVIDataset(data_train, target_col='cell_labels', transform = composed_transforms)
sample = scVI_dataset[0]

In [29]:
sample['features'].dtype

torch.float32

In [30]:
sample['features'].shape

torch.Size([1, 1, 558])

In [31]:
sample['target'].shape

()

In [32]:
sample['target']

5

In [33]:
dataloader = DataLoader(scVI_dataset, batch_size=4, shuffle=False, num_workers=0)
it = iter(dataloader)
s = next(it)

In [34]:
s

{'features': tensor([[[[ 3.3132, -1.0000,  3.5341,  ...,  0.5984, -0.5275,  1.2420]]],
 
 
         [[[-0.4865, -1.0000,  0.0944,  ..., -0.2008,  0.1024,  0.2072]]],
 
 
         [[[ 0.7972, -1.0000, -0.2183,  ...,  0.4386,  0.7324,  0.0348]]],
 
 
         [[[ 0.8485, 41.0596, 12.9149,  ..., -0.3606,  0.7324, -1.0000]]]]),
 'target': tensor([5, 5, 2, 2])}

In [35]:
scVI_dataset[0]

{'features': tensor([[[ 3.3132, -1.0000,  3.5341, -1.0000, -1.0000, -0.8588,  6.9624,
           -0.1789, -1.0000, -1.0000, -0.0368,  0.1388, -0.7543,  5.3429,
            0.5773,  3.5828,  0.2743, -0.3727,  3.2861, -0.2076,  0.9823,
            2.5544, -0.0821, -0.1232,  0.0209,  2.0094,  2.4814,  3.6399,
           -0.2396,  0.0572,  0.6045, -0.8237, -0.7967,  2.0163, -0.0876,
            0.1888,  1.0955,  0.6243,  0.2056,  2.9188, -0.6046,  1.6222,
           -0.4282,  1.0881, -0.2619,  0.1734, -0.2926,  2.0443,  0.6335,
            3.8306, -0.7496,  0.8579,  0.2028,  1.7855,  2.7211,  2.2715,
            1.2406,  1.9087, -1.0000,  1.8797, -1.0000, -1.0000,  1.6393,
            1.2639,  0.9868,  0.4293,  4.4780, -1.0000, -1.0000,  3.2617,
           -1.0000, -1.0000, -1.0000,  0.7918,  1.7557,  0.2209, -0.4443,
           -0.2189, -1.0000, -0.2042, -0.5814, -0.3870, -0.0911,  0.0302,
            0.9874,  0.3432, -0.8171, -0.5404,  0.4560, -0.5202, -0.0356,
           -0.0176,  0.774

In [36]:
#Prepare the train_loader and valid_loader
train_transforms = transforms.Compose([Normalize(norm_mean, norm_std),
                                       ToTensor()])

valid_transforms = transforms.Compose([Normalize(norm_mean, norm_std),
                                       ToTensor()])

train_datasets = ScVIDataset(data_train, target_col='cell_labels', transform = train_transforms)
valid_datasets = ScVIDataset(data_valid, target_col='cell_labels', transform = valid_transforms)

batch_size = 20
num_workers = 0
train_loader = DataLoader(train_datasets, batch_size=batch_size, shuffle=True, num_workers=num_workers)
valid_loader = DataLoader(valid_datasets, batch_size=batch_size, shuffle=True, num_workers=num_workers)
data_loaders = {'train':train_loader, 'valid':valid_loader}

In [37]:
print(f'The number of train datasets is: {len(train_datasets)}')
print(f'The number of valid datasets is: {len(valid_datasets)}')

The number of train datasets is: 2404
The number of valid datasets is: 601


In [38]:
print(f'The number of input features: {len(features)}')
print(f'The number of labels: {len(labels)}')

The number of input features: 558
The number of labels: 7


In [39]:
#implementation of the fcmodel
import torch.nn as nn
import torch.nn.functional as F

#define the architecture of fcmodel: FCNet
class FCNet(nn.Module):
    def __init__(self):
        super(FCNet, self).__init__()
        #define the layers in this model
        #input layer is size len(features)=558
        self.fc1 = nn.Linear(len(features), 256)
        self.fc2 = nn.Linear(256,256)
        self.fc3 = nn.Linear(256,10)
        self.fc4 = nn.Linear(10,len(labels))
        
        self.dropout = nn.Dropout(0.5)
        self.relu = nn.ReLU(inplace=True)
        
    def forward(self, x):
        #Define the forward behavior
        #flat input data first
        x = x.view(x.shape[0], -1)
        x = self.relu(self.fc1(x))
        x = self.dropout(self.relu(self.fc2(x)))
        x = self.dropout(self.relu(self.fc3(x)))
        x = self.fc4(x)
        
        return x        

In [40]:
# the training function
def train(epochs, loaders, model, optimizer, criterion, use_cuda, save_path):
    '''Training function
    Args:
        epochs: the total number of training epochs
        loaders(python dict):with key='train' and 'valid'
        model(): training model
        optimizer:
        criterion:
        use_cuda(boolean): True if cuda is available
        save_path(string of path): the parameters will be saved to the path after training
    '''
    
    valid_loss_min = np.Inf
    
    for epoch in range(1, epochs+1):
        #Initialized to monitor training loss and validatoin loss during training
        train_loss = 0
        valid_loss = 0
        
        # train loop
        #set model to train mode
        model.train()
        for batch_idx, sample in enumerate(loaders['train']):
            data, target = sample['features'], sample['target']
            if use_cuda:
                data, target = data.cuda(), target.cuda()
            optimizer.zero_grad()
            output = model(data)
           
            loss = criterion(output, target)
            loss.backward()
            optimizer.step()
            
            #The loss here is averaged by batch, 
            #  So, using the following equation to calculate the average training_loss during training
            train_loss = train_loss + ((1 / (batch_idx + 1)) * (loss.data - train_loss))
        
        
        # validation loop
        #set the model to evaluation mode
        model.eval()
        #disable auto_grad to speed things up in validation
        with torch.no_grad():
            for batch_idx, sample in enumerate(loaders['valid']):
                data, target = sample['features'], sample['target']
                if use_cuda:
                    data, target = data.cuda(), target.cuda()
                
                output = model(data)
                loss = criterion(output, target)
                valid_loss=valid_loss + ((1 / (batch_idx + 1)) * (loss.data - valid_loss))
                
        #print train loss and valid loss for each epoch
        print(f'Epoch:{epoch} \tTraining Loss:{train_loss:.6f} \tValidation Loss:{valid_loss:.6f}' )
        
        #save the model parameters if validation loss has decreased
        if valid_loss <= valid_loss_min:
            print(f'--Validation loss decreased({valid_loss_min:.6f} --> {valid_loss:.6f})')
            torch.save(model.state_dict(), save_path)
            valid_loss_min = valid_loss
        
    
    return model

In [41]:
#instantiate the model
model_FCNet = FCNet()

is_cuda_available = torch.cuda.is_available()
if is_cuda_available:
    model_FCNet.cuda()

In [42]:
import torch.optim as optim
#setup the optimizer and criterion
optimizer = optim.SGD(model_FCNet.parameters(), lr=0.001, momentum=0.8)

#Here use cross entropy instead of NLLLoss
#  Because there is no softmax at the last layer
##criterion = nn.CrossEntropyLoss()
criterion = nn.CrossEntropyLoss(weight=torch.Tensor(target_weights))

In [43]:
a=torch.Tensor(target_weights)

In [44]:
a.dtype

torch.float32

In [45]:
#clear the cuda cache if using cuda
if is_cuda_available:
    torch.cuda.empty_cache()

model_FCNet_param_save_name = 'model_FCNet2.pt'
#model_train1 = train(100, data_loaders, model_FCNet.double(), optimizer, criterion, is_cuda_available, model_FCNet_param_save_name)
model_train1 = train(100, data_loaders, model_FCNet, optimizer, criterion, is_cuda_available, model_FCNet_param_save_name)

#to offload the GPU
if is_cuda_available:
    torch.cuda.empty_cache()

Epoch:1 	Training Loss:1.936726 	Validation Loss:1.907750
--Validation loss decreased(inf --> 1.907750)
Epoch:2 	Training Loss:1.898881 	Validation Loss:1.869278
--Validation loss decreased(1.907750 --> 1.869278)
Epoch:3 	Training Loss:1.855264 	Validation Loss:1.808286
--Validation loss decreased(1.869278 --> 1.808286)
Epoch:4 	Training Loss:1.792618 	Validation Loss:1.734114
--Validation loss decreased(1.808286 --> 1.734114)
Epoch:5 	Training Loss:1.724917 	Validation Loss:1.645590
--Validation loss decreased(1.734114 --> 1.645590)
Epoch:6 	Training Loss:1.647761 	Validation Loss:1.544606
--Validation loss decreased(1.645590 --> 1.544606)
Epoch:7 	Training Loss:1.551963 	Validation Loss:1.417473
--Validation loss decreased(1.544606 --> 1.417473)
Epoch:8 	Training Loss:1.455385 	Validation Loss:1.257840
--Validation loss decreased(1.417473 --> 1.257840)
Epoch:9 	Training Loss:1.353712 	Validation Loss:1.161474
--Validation loss decreased(1.257840 --> 1.161474)
Epoch:10 	Training Loss:

In [46]:
#load the model parameters with the best validation loss
#model_FCNet.load_state_dict(torch.load(model_FCNet_param_save_name))

In [47]:
#the test function that test the accuracy of the model
def test(test_loader, model, criterion, use_cuda):
    '''Test function
    Args:        
        loaders(python dict):with key='train' and 'valid'
        model(): training model        
        criterion:
        use_cuda(boolean): True if cuda is available        
    '''
    test_loss = 0.0
    correct = 0.0
    total = 0.0
        
    model.eval()
    for batch_idx, sample in enumerate(test_loader):
        data, target = sample['features'], sample['target']
        if use_cuda:
            data, target = data.cuda, target.cuda
        
        output = model(data)
        loss = criterion(output, target)
        test_loss = test_loss + (1/(batch_idx+1)*(loss.data - test_loss))
        
        #compute the probability using softmax
        prob = F.softmax(output, dim=1)
        #find the prediction from the max prob
        pred = prob.data.max(1, keepdim=True)[1]
        #compare predictions and target 
        correct += np.sum(np.squeeze(pred.eq(target.data.view_as(pred))).cpu().numpy())
        total += data.shape[0]
    
    print(f'Loss: {test_loss:.6f}')
    print(f'Accuracy: {100. * correct / total:.2f}%, ({correct}, {total})')
    

In [48]:
#load in the previous training result
model_test_name = 'model_FCNet2.pt'
print(f'Loaded model parameters from: {model_test_name}')
model_FCNet.load_state_dict(torch.load(model_test_name))

Loaded model parameters from: model_FCNet2.pt


In [49]:
print('Loss and Accuracy of test set')
test(data_loaders['valid'], model_FCNet, criterion, is_cuda_available)

Loss and Accuracy of test set
Loss: 0.166207
Accuracy: 96.34%, (579.0, 601.0)


In [50]:
print('Loss and Accuracy of train set')
test(data_loaders['train'], model_FCNet, criterion, is_cuda_available)

Loss and Accuracy of train set
Loss: 0.048267
Accuracy: 99.54%, (2393.0, 2404.0)


# Result Notes:

`lr` = 0.001, `momentum` = 0.9: file: __model_FCNet1.pt__
0. results: 
    Test Loss: 0.333796<br>
    Test Accuracy: *93.34%, (561.0, 601.0)* <br>
    Train Loss: 0.166586<br>
    Train Accuracy: *95.97%, (2307.0, 2404.0)* <br>
1. no weight in CrossEntropyLoss
    

`lr` = 0.001, `momentum` = 0.8: file: __model_FCNet2.pt__
0. results:
    Test Loss: 0.166207<br>
    Test Accuracy: __96.34%__, (579.0, 601.0)<br>
    Train Loss: 0.048267<br>
    Train Accuracy: __99.54%__, (2393.0, 2404.0)<br>
1. with weight in CrossEntropyLoss

`lr` = ????, `momentum` = ????: file: __model_FCNet3.pt__
